compiler/rustc_abi/src/canon_abi.rs

```diff
@@ -0,0 +1,237 @@
+    // NOTE: the use of nested variants for some ABIs is for many targets they don't matter,
+    // and this pushes the complexity of their reasoning to target-specific code,
+    // allowing a `match` to easily exhaustively ignore these subcategories of variants.
+    // Otherwise it is very tempting to avoid matching exhaustively!
+    C,
+    Rust,
+    RustCold,
+
+    /// ABIs relevant to 32-bit Arm targets
+    Arm(ArmCall),
+    /// ABI relevant to GPUs: the entry point for a GPU kernel
+    GpuKernel,
+
+    /// ABIs relevant to bare-metal interrupt targets
+    // FIXME(workingjubilee): a particular reason for this nesting is we might not need these?
+    // interrupt ABIs should have the same properties:
+    // - uncallable by Rust calls, as LLVM rejects it in most cases
+    // - uses a preserve-all-registers *callee* convention
+    // - should always return `-> !` (effectively... it can't use normal `ret`)
+    // what differs between targets is
+    // - allowed arguments: x86 differs slightly, having 2-3 arguments which are handled magically
+    // - may need special prologues/epilogues for some interrupts, without affecting "call ABI"
+    Interrupt(InterruptKind),
+
+    /// ABIs relevant to Windows or x86 targets
+    X86(X86Call),
+}
+
+// For most "utility" impls, just forward CanonAbi to ExternAbi as it is notionally a subset
+impl Ord for CanonAbi {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.to_erased_extern_abi().cmp(&other.to_erased_extern_abi())
+    }
+}
+
+impl PartialOrd for CanonAbi {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl PartialEq for CanonAbi {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(&other) == Ordering::Equal
+    }
+}
+
+impl Eq for CanonAbi {}
+
+impl Hash for CanonAbi {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.to_erased_extern_abi().hash(state)
+    }
+}
+
+// because we forward hashing to ExternAbi, also forward this:
+#[cfg(feature = "nightly")]
+impl<C> HashStable<C> for CanonAbi {
+    #[inline]
+    fn hash_stable(&self, _: &mut C, hasher: &mut StableHasher) {
+        Hash::hash(self, hasher);
+    }
+}
+
+#[cfg(feature = "nightly")]
+impl StableOrd for CanonAbi {
+    const CAN_USE_UNSTABLE_SORT: bool = true;
+
+    // because each ABI is hashed like a string, there is no possible instability
+    const THIS_IMPLEMENTATION_HAS_BEEN_TRIPLE_CHECKED: () = ();
+}
+
+impl fmt::Display for CanonAbi {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.to_erased_extern_abi().fmt(f)
+    }
+}
+
+impl CanonAbi {
+    /// convert to the ExternAbi that *shares a string* with this CanonAbi
+    ///
+    /// NOT correct to use if you want to map CanonAbi to an ABI it may have been lowered from,
+    /// but it is convenient for various "forwarding" implementations. Avoid exposing publicly!
+    const fn to_erased_extern_abi(self) -> ExternAbi {
```

Comment:
It seems to be important that this is injective, right? That should be spelled out.